### ISSUE TITLE
Potential race condition in closures

### ISSUE THREAD

**Issue by krugazor** [IBM-Swift/Swift-Kuery-PostgreSQL#23]
*Context*: a very large postgresql db (> 100k hits per query), a function that queries for the existence of 2 conditions, then inserts data if the conditions aren't met.

*Expected result*: no error, insertion

*Actual result*: no error (TUPLE_OK from the pq lib, successNoData from Kuery), no insertion

*More details*:

the basic layout of the code was:
```
query { result in
  query { result in {
    insert
  }
}
```

What did work was putting the insert on a separate connection from the pool. 

```
let conn = pool.getConnection() 
conn.query { result in {
  conn.query { result in {
    let conn2 = pool.getConnection()
    conn2.insert
  }
}
```

After a conversation with @shmuelk, I opened this issue, but I'm aware of the difficulties reproducing such a bug with access to my exact setup.

**Comment by LoiKos** [comment]
Hi,

I reach a similar mistake and after a few research, there is my conclusion : 

When you are sending the request you receive an `QueryResult`object .  This query result contains Error or ResultSet or nothing. If you try chaining request as you show, you will end up with resultNoData as QueryResult. I can't understand why but it's like that. 

Know if you do the same with a little bit of parsing : 

```swift 
query { result in
switch (result) {
 case .resultSet:
   query { result in {
    
   }
case default : throw error 
}```

There ( for me at least) I can access resulSet  correctly

**Comment by irar2** [comment]
The problem here is that when doing

```
query1 { result1 in
  query2 { result2 in {
    insert
  }
}
```

query2 is executed when the connection is in the middle of reading the results of query1, i.e. the call to PQgetResult(connection) still gets the next row of query1 even in the second callback for result2.asResultSet.nextRow().

We have to decide what to do in such situations, when we are in the middle of fetching results and receive an 'execute' call with another query. We can either  kill the first query fetcher, or don't execute the second query and return an error for it.

The branch issue_23 contains the first option. It adds a state to `PostgreSQLConnection`, which is checked before query execution, and kills the existing fetcher in case the state is `.fetchingResultSet`.

**Comment by djones6** [comment]
The fix has been merged in #45.
